File: code_snippets/ts/src/develop/my_virtual_object.ts
import * as restate from "@restatedev/restate-sdk";
import {handlers, ObjectContext, ObjectSharedContext} from "@restatedev/restate-sdk";

const myVirtualObject = restate.object({
    name: "MyVirtualObject",
    handlers: {
        myHandler: async (ctx: ObjectContext, greeting: string) => {
            return `${greeting} ${ctx.key}!`;
        },
        myConcurrentHandler: handlers.object.shared(
            async (ctx: ObjectSharedContext, greeting: string) => {
                return `${greeting} ${ctx.key}!`;
            }
        ),
    }
})

export const MyVirtualObject: typeof myVirtualObject = { name: "MyVirtualObject" };

restate
    .endpoint()
    .bind(myVirtualObject)
    .listen();

File: code_snippets/python/src/develop/my_virtual_object.py
import restate
from restate import VirtualObject, ObjectContext, ObjectSharedContext

my_object = VirtualObject("MyVirtualObject")


@my_object.handler("myHandler")
async def my_handler(ctx: ObjectContext, greeting: str) -> str:
    return f"${greeting} ${ctx.key()}!"


@my_object.handler(kind="shared")
async def my_concurrent_handler(ctx: ObjectSharedContext, greeting: str) -> str:
    return f"${greeting} ${ctx.key()}!"


app = restate.app([my_object])
